### PR TITLE
TTL Indexes

### PR DESCRIPTION
**Pull request by juannyG**
Implement https://github.com/mongomock/mongomock/issues/655

Key points:

1. The first thing the `_iter_documents` method does now is remove expired documents. From what I could tell, this seemed to be the central point where any and all document examination occurred.
2. The TTL "time frame-of-reference" was made mock-able by using a centralized wrapper of `datetime.datetime.utcnow` - however further centralization can be handled separately. Given the stated desire, I wanted to try and plant this seed.

There are some other could improvements I can/would like to eventually add, specifically, some kind of `IndexWrapper` object to drastically improve the readability of the `_expire_documents` method, for example:

`idx.is_compound()` instead of `if len(index['key']) > 1`

but I wanted to get the core of the feature and where it might live in front of you first.

Feature and tests driven by: https://docs.mongodb.com/manual/core/index-ttl/

### CODE REVIEW

File: mongomock/helpers.py [hunk 1]
```diff
@@ -49,6 +49,10 @@ def dst(self, dt):
 DESCENDING = -1
 
 
+def utcnow():
```
Comment (pcorpet) [inline mongomock/helpers.py:52]: Please document: what it does and why we handle it here

File: mongomock/store.py [hunk 2]
```diff
@@ -104,3 +107,51 @@ def __len__(self):
     def documents(self):
         for doc in six.itervalues(self._documents):
             yield doc
+
+    def remove_expired_documents(self):
```
Comment (pcorpet) [inline mongomock/store.py]: I would prefer if the responsibility of never returning expired document would land in the store class. Therefore I believe this method should be protected and called in the other methods in this class.

Two reasons why I prefer it to be the responsibility of the store rather than the collection:
* the store is mocking the server's storage, whereas the collection is the client's view of it and the expiration of documents is done in the server
* there is no reasons ever that the store should returned expired document

File: mongomock/store.py [hunk 3]
```diff
@@ -104,3 +107,51 @@ def __len__(self):
...
+        for index in six.itervalues(self.indexes):
+            if index.get('expireAfterSeconds') is not None:
```
Comment (pcorpet) [inline mongomock/store.py]: Nit - you could create a set of indexes that have an expireAfterSeconds field instead of checking them each time. I'm trying to optimize because this method is probably going to be called many many times.

File: mongomock/store.py [hunk 4]
```diff
@@ -104,3 +107,51 @@ def __len__(self):
...
+                self._expire_documents(index)
+
+    def _expire_documents(self, index):
+        # Ignore non-integer values
```
Comment (pcorpet) [inline mongomock/store.py:139]: Please add a TODO to use a caching mechanism to avoid re-expiring the documents if we just did and that no document was added / updated.

File: mongomock/store.py [hunk 5]
```diff
@@ -104,3 +107,51 @@ def __len__(self):
...
+        except ValueError:
+            return
+
+        # Ignore commpound keys
+        if len(index['key']) > 1:
+            return
+
+        # "key" structure = list of (field name, direction) tuples
+        ttl_field_name = index['key'][0][0]
+        expired_ids = {
```
Comment (pcorpet) [inline mongomock/store.py]: I'm not sure doc['_id'] have to be hashable, so I suggest you build a list instead of building a set.

File: mongomock/store.py [hunk 6]
```diff
@@ -104,3 +107,51 @@ def __len__(self):
...
+
+    def _value_meets_expiry(self, val, expiry):
+        val_to_compare = _get_dt_from_value(val)
+        try:
+            return (mongomock.utcnow() - val_to_compare).total_seconds() >= expiry
+        except TypeError:
+            return False
+
+
+def _get_dt_from_value(val):
```
Comment (pcorpet) [inline mongomock/store.py]: please rename this function to `_get_min_datetime_from_value` or just `_get_min_datetime`

File: mongomock/helpers.py (hunk 1 above)
Comment (pcorpet) [inline mongomock/helpers.py:52]: Please update `_current_date_updater` to use this

File: mongomock/store.py [hunk 7]
```diff
@@ -104,3 +107,51 @@ def __len__(self):
...
+            if self._value_meets_expiry(doc.get(ttl_field_name), expiry)
+        }
+
+        for exp_id in expired_ids:
+            del self[exp_id]
+
+    def _value_meets_expiry(self, val, expiry):
+        val_to_compare = _get_dt_from_value(val)
+        try:
+            return (mongomock.utcnow() - val_to_compare).total_seconds() >= expiry
```
Comment (pcorpet) [inline mongomock/store.py]: Instead of calling utcnow on each document, I suggest you call utcnow only once in _expire_documents, or even in remove_expired_documents.

File: tests/test__collection_api.py [hunk 8]
```diff
@@ -1313,9 +1323,90 @@ def test__create_index_wrong_type(self):
         with self.assertRaises(TypeError):
             self.db.collection.create_index([('value', 1, 'foo', 'bar')])
 
-    def test__create_ttl_index(self):
-        with self.assertRaises(NotImplementedError):
-            self.db.collection.create_index([('value', 1)], expireAfterSeconds=3600)
+    def test__ttl_index_record_expiration(self):
```
Comment (pcorpet) [inline tests/test__collection_api.py]: Please split in 3 test functions (as you can see, we are happy having multiple test functions, and here you might run into trouble because one test is leaking into the other, as your `drop` call suggests)

**Review by pcorpet — CHANGES_REQUESTED** [review]
Wow, you made it! Please take a look at my comments.

File: mongomock/store.py (hunk 2 above)
Comment (juannyG) [reply mongomock/store.py]: Gotcha. I had thought of that originally, but once I saw that basically every method of the store would need to call this method, I started lurking elsewhere :wink:

File: mongomock/store.py (hunk 3 above)
Comment (juannyG) [reply mongomock/store.py]: I added some index helper methods to the store to ensure accurate state tracking while also satisfying this optimization.

**Comment by juannyG** [comment]
Not sure what's up with the `pylint` run on that last build - locally I get 10.00/10

:thinking:

File: mongomock/collection.py [hunk 9]
```diff
@@ -2050,7 +2052,7 @@ def _current_date_updater(doc, field_name, value):
         if value == {'$type': 'timestamp'}:
             doc[field_name] = helpers.get_current_timestamp()
```
Comment (pcorpet) [inline mongomock/collection.py:2056]: Note that we should also update this one somehow. Please add a TODO or find a way already.

File: mongomock/collection.py [hunk 10]
```diff
@@ -2050,7 +2052,7 @@ def _current_date_updater(doc, field_name, value):
...
         else:
-            doc[field_name] = datetime.utcnow()
+            doc[field_name] = mongomock.utcnow()
```
Comment (pcorpet) [inline mongomock/collection.py:2058]: Please import it from helpers instead of mongomock

File: mongomock/store.py [hunk 11]
```diff
@@ -78,29 +82,97 @@ def is_created(self):
...
+        except ValueError:
+            return
+
+        # Ignore commpound keys
+        if len(index['key']) > 1:
+            return
+
+        # "key" structure = list of (field name, direction) tuples
+        ttl_field_name = index['key'][0][0]
+        ttl_now = mongomock.utcnow()
```
Comment (pcorpet) [inline mongomock/store.py:151]: Here as well, import it from helpers please.

File: mongomock/store.py [hunk 12]
```diff
@@ -78,29 +82,97 @@ def is_created(self):
...
+        self._remove_expired_documents()
         return self._documents[key]
 
     def __setitem__(self, key, val):
         with lock:
             self._documents[key] = val
 
     def __delitem__(self, key):
-        del self._documents[key]
+        self._documents.pop(key)
```
Comment (pcorpet) [inline mongomock/store.py]: why the change here? I find the `del` version more readable.

Also, shouldn't you need to remove expired documents before removing the key? I guess we are missing tests around deleting non existing documents.

**Review by pcorpet — APPROVED** [review]

File: mongomock/store.py (hunk 11 above)
Comment (juannyG) [reply mongomock/store.py:151]: I can do this but it has an interesting side-effect - [the test specific to mocking will fail](https://github.com/juannyG/mongomock/blob/ttl-index/tests/test__collection_api.py#L1389) because the "usage path" has changed.

There are two ways to go about this - the first I think will be less appealing, because I believe it goes against your original intent.

In `monogmock/store.py`

    from mongomock import helpers
    ...
    ttl_now = helpers.utcnow()

the test patch would then become

    with mock.patch('mongomock.store.helpers.utcnow') as mongomock_utcnow:

The implication of this is that now users would have to be aware of the different usages/locations in the library to target appropriately.

The alternative would allow for centralized mock that would affect all usages in the library that follow the same usage-pattern:

In `monogmock/store.py`

    import mongomock.helpers
    ...
    ttl_now = mongomock.helpers.utcnow()

the test patch would then become

    with mock.patch('mongomock.helpers.utcnow') as mongomock_utcnow:

From your earlier comments in the issue I originally opened, it seems as though the latter is the preferred approach: as long as anywhere in the library we want to allow mocking access uses `helpers.utcnow` in the same way - users would only need to mock in one place for "global" `utcnow` manipulation.

Subtle - but I wanted to raise this and get your input before I made any more changes. Thx!

File: mongomock/collection.py (hunk 9 above)
Comment (juannyG) [reply mongomock/collection.py:2056]: I'll try and revisit this separately

**Comment by juannyG** [comment]
Separately - I was able to replicate the `pylint` errors travis is reporting - I needed to run `tox -r` to get a fresh tox env...but they don't seem related to my changes.

**Comment by juannyG** [comment]
Digging deeper - using [six.raise_from](https://six.readthedocs.io/#six.raise_from) appears to clear up the pylint issues and appears backwards compatible (tests still pass).

LMK if you'd like for me to include those changes in this PR to keep the build nice and green :smile:

File: mongomock/store.py (hunk 11 above)
Comment (pcorpet) [reply mongomock/store.py:151]: I see your point, and you're entirely right. Let's just use `mongomock.utcnow` in our code as well. I'll try to make sure we don't use the `mongomock` import internally for other reasons. Maybe add a comment just above the `import mongmomock` explaining why we do it like that? Maybe just redirect readers to the README (no need for a huge explanation).

**Comment by pcorpet** [comment]
I'll fix pylint errors at HEAD of develop.

**Comment by juannyG** [comment]
Let me know if there is there anything else you'd like addressed or aren't sure about with these changes - thx!
